## Passwords Competition

In [12]:
import numpy as np
import pandas as pd

In [13]:
sample = pd.read_csv('/media/xenakas/ext4_drive/Kaggle/DMIA/passwords/sample_submission.csv') 
train =  pd.read_csv('/media/xenakas/ext4_drive/Kaggle/DMIA/passwords/train.csv') 
test =   pd.read_csv('/media/xenakas/ext4_drive/Kaggle/DMIA/passwords/Xtest.csv', index_col=0) 

/home/xenakas/anaconda3/lib/python3.6/site-packages/numpy/lib/arraysetops.py:571: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


In [14]:
count_train = train.groupby('Times').count()
count_train.head(10)

Password
Times          
1       3547155
2        309205
3         96461
4         47704
5         27903
6         18627
7         13635
8         10567
9          8235
10         6775

In [15]:
top_50 = np.array(count_train.iloc[-50:, 0].index)
top_50

array([ 1635,  1648,  1664,  1677,  1688,  1692,  1699,  1702,  1705,
        1718,  1727,  1771,  1783,  1861,  1868,  1901,  1935,  1964,
        2045,  2074,  2085,  2155,  2189,  2221,  2296,  2315,  2340,
        2345,  2480,  2531,  2573,  2665,  2804,  2860,  2905,  2931,
        2956,  3118,  3494,  3511,  3565,  3845,  3927,  4796,  5682,
        6432, 10938, 11696, 13137, 55893])

In [16]:
train[train['Times'].isin(top_50)]

Password  Times
49123      computer   1688
177871       shadow   2956
311449       hunter   2045
385137     1qaz2wsx   2531
425279      fuckyou   2345
553802      asshole   1692
584867      mustang   2804
625963       123qwe   2296
712456       killer   2221
769528      zxcvbnm   2085
804594       hockey   1783
1120418     7777777   2480
1129735      george   1699
1166785      ranger   1771
1175081   123456789  11696
1525092      112233   1702
1644230      fuckme   1868
1703126      zxcvbn   1635
1747604     letmein   3118
1795496      dragon   3927
1807658    jennifer   2155
1864994    sunshine   1901
2089878      asdfgh   2074
2177478     jessica   1664
2257222       pussy   2573
2307329      111111   5682
2363307       12345  10938
2490203    michelle   1677
2628304      jordan   2189
2678990      master   2931
2715397      123456  55893
2788383     klaster   1705
2845798      666666   2905
2988373        1234   6432
3061106  qwertyuiop   2860
3096637      654321   2665
3135141      abc123   3511
3136279      qwerty  13137
3229701      121212   2340
3302574        2000   1861
3336606      123123   3845
3348280     1234567   4796
3351563    starwars   1718
3451509        1111   1648
3541881    football   3494
3783021      batman   1935
3798071    baseball   3565
3804747      soccer   1964
3834388      daniel   1727
3883643      qazwsx   2315

### Creating features

#### Feature 1

In [65]:
def feature_1(inputString):
    valid_symbols = "!?.,;:~`'<>/\|_@#$%^&*()_-+={}[]"    
    if inputString.isalpha(): 
        return 1 # only letters
    elif inputString.isdigit(): 
        return 2 # only numbers
    elif inputString.isalnum():
        return 3 # letters + numbers
    elif all(char in valid_symbols for char in inputString): 
        return 7 # only symbols
    elif (any(char in valid_symbols for char in inputString)) & (any(char.isalpha() for char in inputString)) & (any(char.isdigit() for char in inputString)): 
        return 4 # letters + numbers + symbols
    elif (any(char in valid_symbols for char in inputString)) & (any(char.isalpha() for char in inputString)): 
        return 5 # letters + symbols
    elif (any(char in valid_symbols for char in inputString)) & (any(char.isdigit() for char in inputString)):
        return 6 # numbers + symbols
    else: 
        return 8 

    

In [54]:
train.Password = train.Password.apply(str)

In [66]:
train['Feature_1']  = train.Password.apply(feature_1)

In [75]:
train.groupby('Feature_1').sum()

Times
Feature_1         
1          3331255
2          1624691
3          2935279
4            42728
5            38619
6             6636
7              486
8               32

In [80]:
f11 = train[train['Feature_1']==1 ]
f12 = train[train['Feature_1']==2 ]
f13 = train[train['Feature_1']==3 ]
f14 = train[train['Feature_1']==4 ]
f15 = train[train['Feature_1']==5 ]
f16 = train[train['Feature_1']==6 ]
f17 = train[train['Feature_1']==7 ]

In [97]:
count11 = f11.groupby('Times').count()
top11 = np.array(count11.iloc[-50:, 0].index)
train[train['Times'].isin(top11)]

Password  Times  Feature_1
49123      computer   1688          1
177871       shadow   2956          1
311449       hunter   2045          1
392514         pass   1550          1
425279      fuckyou   2345          1
467757       ashley   1428          1
553802      asshole   1692          1
584867      mustang   2804          1
662284      matthew   1369          1
712456       killer   2221          1
769528      zxcvbnm   2085          1
784349       nicole   1393          1
804594       hockey   1783          1
928280       ginger   1491          1
1011479        fuck   1520          1
1129735      george   1699          1
1166785      ranger   1771          1
1179075      dallas   1344          1
1183989      amanda   1462          1
1503202    princess   1483          1
1538061     yankees   1361          1
1552465      cheese   1465          1
1644230      fuckme   1868          1
1703126      zxcvbn   1635          1
1747604     letmein   3118          1
1795496      dragon   3927          1
1807658    jennifer   2155          1
1864994    sunshine   1901          1
2089878      asdfgh   2074          1
2177478     jessica   1664          1
2257222       pussy   2573          1
2414860      summer   1444          1
2490203    michelle   1677          1
2628304      jordan   2189          1
2678990      master   2931          1
2788383     klaster   1705          1
2826330        love   1433          1
3061106  qwertyuiop   2860          1
3136279      qwerty  13137          1
3295554     freedom   1563          1
3302432      joshua   1483          1
3351563    starwars   1718          1
3541881    football   3494          1
3542490      maggie   1495          1
3783021      batman   1935          1
3798071    baseball   3565          1
3804747      soccer   1964          1
3813535      biteme   1376          1
3834388      daniel   1727          1
3857521      aaaaaa   1492          1
3883643      qazwsx   2315          1

In [99]:
count12 = f12.groupby('Times').count()
top12 = np.array(count12.iloc[-50:, 0].index)
train[(train['Times'].isin(top12))&(train['Feature_1']==2)]

Password  Times  Feature_1
97244        333333    910          2
186734     12341234    717          2
495409         1990    733          2
597598         1991    720          2
608073       555555   1591          2
611931    987654321   1349          2
720813       101010    726          2
850685       159357    739          2
1120418     7777777   2480          2
1175081   123456789  11696          2
1320706    11111111   1575          2
1525092      112233   1702          2
1551210      007007    540          2
1699163   123123123   1016          2
1755942     8675309    801          2
1854288        5150    626          2
2158093      222222   1246          2
2307329      111111   5682          2
2363307       12345  10938          2
2462772    69696969    583          2
2573476      888888    826          2
2619424    88888888   1243          2
2715397      123456  55893          2
2823562        1989    613          2
2845798      666666   2905          2
2854521       00000    541          2
2905714        1992    709          2
2988373        1234   6432          2
3096637      654321   2665          2
3134420   789456123    545          2
3135254    01012011    586          2
3201939  4815162342    607          2
3229701      121212   2340          2
3268986      212121    616          2
3294349      444444    652          2
3302574        2000   1861          2
3336606      123123   3845          2
3348280     1234567   4796          2
3450338      315475    574          2
3451509        1111   1648          2
3464277      123654    986          2
3500832      252525    578          2
3641387        6969   1426          2
3667067     1111111    632          2
3765524       11111   1186          2
3800937        7777    642          2
3865480      999999    958          2
3886656        0000    944          2
4050598    12344321    808          2
4103669        1988    564          2
4108956        1993    564          2

In [101]:
count13 = f13.groupby('Times').count()
top13 = np.array(count13.iloc[-50:, 0].index)
train[train['Times'].isin(top13)&(train['Feature_1']==3)]

Password  Times  Feature_1
364291       bond007    636          3
378651        12345a    615          3
385137      1qaz2wsx   2531          3
435242      jordan23    776          3
549179     qwerty123    722          3
580631    12345qwert    434          3
625963        123qwe   2296          3
639369        eagle1    445          3
641249        a1b2c3    339          3
819320      asdf1234    518          3
829341       ncc1701    956          3
873239        1q2w3e    349          3
1158668     blink182    553          3
1184551       asd123    480          3
1303854     ncc1701d    374          3
1398104      qwerty1    472          3
1419265   q1w2e3r4t5   1201          3
1454040       qaz123    330          3
1602314     hello123    366          3
1829035     rush2112    613          3
1915023     1234qwer   1276          3
2076258     12qwaszx    543          3
2145857     michael1    344          3
2179171     zaq12wsx    397          3
2250529        ou812    475          3
2288702      123456q    646          3
2333849       qqq111    339          3
2342532  Usuckballz1    417          3
2376398   123456789q    336          3
2440643       q1w2e3    646          3
2485367       lol123    503          3
2538635       a12345    422          3
2737163      rasdzv3    376          3
2740198       red123    608          3
2781093       qwe123    430          3
2829792       hello1    456          3
2881720      123456a    733          3
2954574     1q2w3e4r    846          3
2992310   123456789a    420          3
3051133       lucky1    423          3
3131093       12345q    468          3
3135141       abc123   3511          3
3186333      thx1138    790          3
3298921       123abc    757          3
3402675     q1w2e3r4    933          3
3438682    password1    723          3
3631321     passw0rd    606          3
3777826      a123456    418          3
3833644     1qazxsw2    590          3
3846946     qwer1234    997          3
3906753   1q2w3e4r5t    647          3
4022058     1232323q    450          3

In [113]:
count14 = f14.groupby('Times').count()
top14 = np.array(count14.iloc[-15:, 0].index)
train[train['Times'].isin(top14)&(train['Feature_1']==4)]

Password  Times  Feature_1
1162          ZAQ!2wsx     15          4
224218    diunilaobu8*      9          4
360200        tr@nsf3r      8          4
690722         abc_123      9          4
702825       ybrbnf_25     15          4
731077         keli_14     18          4
815678    P030710P$E4O     17          4
818757        doc_0815     12          4
833935       zgmf-x10a      8          4
1207917          1-Oct     15          4
1216864      %E2%82%AC     13          4
1383298      sasha_007     74          4
1621192   g00dPa$$w0rD     11          4
1750702          2-Oct     11          4
1839809         4_LiFe     24          4
1923056       !QAZ2wsx     37          4
2027122    h_froeschl7      9          4
2210893       s69!#%&(      9          4
2212383      fre_ak8yj     13          4
2280281    Jhon@ta2011     10          4
2367174       za1986__      8          4
2595752        str|ct9     11          4
2916284    Nloq_010101      9          4
2996639  xxPa33bq.aDNA     11          4
3048334       !QAZxsw2     15          4
3125370     4rdf_king7     11          4
3171028       1qaz@WSX     17          4
3195829       P@ssw0rd     44          4
3203083       wapbbs_1     12          4
3261542       !QAZ1qaz     11          4
3301603        p@ssw0r      8          4
3426359       1qaz!QAZ     37          4
3572495          8-Oct      8          4
3721850       p@ssw0rd     25          4
4102433      L58jkdjP!     46          4

In [112]:
count15 = f15.groupby('Times').count()
top15 = np.array(count15.iloc[-25:, 0].index)
train[train['Times'].isin(top15)&(train['Feature_1']==5)]

Password  Times  Feature_1
46922          rjvgm.nth     18          5
138523          fuckyou!     20          5
155699           n.kmgfy     16          5
306859            tony_t     17          5
336888              poo_     16          5
384109             k.jdm     40          5
532385          mor_pass     12          5
629221          asdfjkl;     41          5
719562           .ktymrf     25          5
725003       fuck_inside    165          5
766158           k.ljxrf     11          5
996603           cool-ca     13          5
1004548        .adgjmptw     21          5
1478706           .adgjm     24          5
1479454           .kbxrf     11          5
1513088          ytgjvy.     11          5
1646160        rc.itymrf     11          5
1657114       bot_schokk     11          5
1776806         fy.njxrf     17          5
1919068          rfn.irf     27          5
1921898            (null     15          5
2044936  ne_e_pod_chehyl     13          5
2260306          yfl.irf     16          5
2296173           irock.     16          5
2327182           ****me     20          5
2332991          hip-hop     20          5
2495000           .kzirf     14          5
2630895         xcat_xca     14          5
2656057         vovan_lt     12          5
2686233           pic\'s     55          5
2759819          k.lvbkf     69          5
2844084         is_a_bot     18          5
2971223             f**k    104          5
2979809         fylh.irf     17          5
2984023      FUCK_INSIDE     25          5
3038069           #NAME?     11          5
3094568         zxcvbnm.     22          5
3125074          orkiox.     34          5
3350498           fy.nrf     35          5
3446613          my_pass     14          5
3641245           .hjxrf     12          5
3696307     minnesota_hp     14          5
3707589           rc.irf     11          5
3712951         m_roesel     13          5
3913353        iloveyou!     29          5
4075519         WAS.HERE     13          5
4077790         %%passwo     35          5
4086561           bk.irf     11          5
4115446           w_pass     73          5
4135208           .ktxrf     26          5

In [119]:
count16 = f16.groupby('Times').count()
top16 = np.array(count16.iloc[-10:, 0].index)
train[train['Times'].isin(top16)&(train['Feature_1']==6)]

Password  Times  Feature_1
337942       270873_     30          6
482128       123456-     11          6
486295        12345$     17          6
547445       123-456      8          6
913944      2004-10-     19          6
1188445     2004-08-      8          6
1778206      123456_      8          6
1782136      3.14159      8          6
2562439  1234567890-     60          6
2655167   123456789.     18          6
2676208    0.0.0.000    131          6
3010867   123456789*     11          6
3254946       13579-     35          6
3447189     2004-11-     13          6

In [123]:
count17 = f17.groupby('Times').count()
top17 = np.array(count17.iloc[-10:, 0].index)
train[train['Times'].isin(top17)&(train['Feature_1']==7)]

Password  Times  Feature_1
54785     ???????     16          7
348196       &*%#      9          7
462662      _____      7          7
1088271     *****    100          7
1278142   *******     23          7
2115909     ?????     65          7
2469139  ********     21          7
2611764    @@@@@@      8          7
2964453    ??????     89          7
3098420    !!!!!!      8          7
3500624      ****     51          7
3762358    !@#$%^      7          7

#### Feature 2

- is substring of "~!@#$%^&*()_+" or "qwertyuiop" or "asdfghjkl" or "zxcvbnm" or "01234567890"

#### Feature 3

- all characters are the same

#### Feature 4

- looks like date (or has 19**)

#### Feature 5 

- is a word

#### Feature 6
- is "p\*\*\*w\*rd"